### PR TITLE
Feature/teams hook

### PR DESCRIPTION
**Pull request by matthdsm**
This PR adds a config option + function to enable messaging pipeline status updates through Teams

Examples:

Collapsed

Opened

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [ ] `CHANGELOG.md` is updated
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] Documentation in `docs` is updated

### CODE REVIEW

**Comment by matthdsm** [comment]
I haven't updated the changelog yet since there're no new entry available

File: nf_core/pipeline-template/lib/NfcoreTemplate.groovy [hunk 1]
```diff
@@ -145,6 +145,61 @@ class NfcoreTemplate {
         output_tf.withWriter { w -> w << email_txt }
     }
 
+    //
+    // Construct and send adaptive card
+    // https://adaptivecards.io
+    //
+    public static void adaptivecard(workflow, params, summary_params, projectDir, log, multiqc_report=[]) {
```
Comment (mirpedrol) [inline nf_core/pipeline-template/lib/NfcoreTemplate.groovy:152]: `log` and `multiqc_report` are not used, I think they can be removed, or even better add a log message if something fails.

File: nf_core/pipeline-template/lib/NfcoreTemplate.groovy [hunk 2]
```diff
@@ -145,6 +145,61 @@ class NfcoreTemplate {
...
+
+        // POST
+        def post = new URL(hook_url).openConnection();
+        post.setRequestMethod("POST")
+        post.setDoOutput(true)
+        post.setRequestProperty("Content-Type", "application/json")
+        post.getOutputStream().write(json_message.getBytes("UTF-8"));
+        def postRC = post.getResponseCode();
+        if (! postRC.equals(200)) {
+            println(post.getErrorStream().getText());
```
Comment (mirpedrol) [inline nf_core/pipeline-template/lib/NfcoreTemplate.groovy:199]: log could be used here

**Review by mirpedrol — COMMENTED** [review]
It looks good to me! But maybe someone with more Groovy knowledge can have a better look 🙂

File: nf_core/pipeline-template/lib/NfcoreTemplate.groovy (hunk 2 above)
Comment (matthdsm) [inline nf_core/pipeline-template/lib/NfcoreTemplate.groovy:199]:
```suggestion
            log.warn(post.getErrorStream().getText());
```

Comment (matthdsm) [reply nf_core/pipeline-template/lib/NfcoreTemplate.groovy:199]: This to your liking @mirpedrol ?

**Comment by matthdsm** [comment]
I'm aware there's a lot of duplicate code with the e-mail function, but I'm not really sure how I would go about cleaning that up in a "groovy" way.

File: nf_core/pipeline-template/lib/NfcoreTemplate.groovy (hunk 1 above)
Comment (matthdsm) [reply nf_core/pipeline-template/lib/NfcoreTemplate.groovy:152]:
```suggestion
    public static void adaptivecard(workflow, params, summary_params, projectDir, log) {
```

File: nf_core/pipeline-template/lib/NfcoreTemplate.groovy (hunk 2 above)
Comment (mirpedrol) [reply nf_core/pipeline-template/lib/NfcoreTemplate.groovy:199]: Looks good!

**Comment by matthdsm** [comment]
Perhaps groovy master @mahesh-panchal would care to comment on the groovy part?

**Comment by mirpedrol** [comment]
> I'm aware there's a lot of duplicate code with the e-mail function, but I'm not really sure how I would go about cleaning that up in a "groovy" way.

I think the idea is to let Nextflow handle email (see #112), so it's probably not worth it to put a lot of effort into cleaning this.

File: nf_core/pipeline-template/workflows/pipeline.nf [hunk 3]
```diff
@@ -114,6 +114,9 @@ workflow.onComplete {
         NfcoreTemplate.email(workflow, params, summary_params, projectDir, log, multiqc_report)
     }
     NfcoreTemplate.summary(workflow, params, log)
+    if (params.hook_url) {
+        NfcoreTemplate.adaptivecard(workflow, params, summary_params, projectDir, log, multiqc_report)
```
Comment (matthdsm) [inline nf_core/pipeline-template/workflows/pipeline.nf:118]:
```suggestion
        NfcoreTemplate.adaptivecard(workflow, params, summary_params, projectDir, log)
```

**Review by mirpedrol — APPROVED** [review]
Checking on that again. Let's merge it and test it on dev if you are happy with it 🙂 
Remember to update the CHANGELOG before

**Comment by matthdsm** [comment]
Updated the changelog!

**Comment by mirpedrol** [comment]
@nf-core-bot fix linting

**Pull request merged**